### PR TITLE
Clean machine id when restored with null context

### LINKED ISSUE #390: resetStateMachine does not clear machine id relating to prior state context
In AbstractStateMachine.java the resetStateMachine method does not clear all machine fields upon being called with a null context (ie a new instance of StateContext to run in the machine).

Observations are that only currentState and extendedState are updated / cleared as part of the operation, I'd suggest that clearing the machine id should be included here to avoid accidentally using a stale id which refers to a context that is no longer loaded in the machine, but it may make sense to clear our other member fields also.

(I say set to null as I'm not sure it's the correct place to accept a new ID as part of method signature along with the context, but it might be a useful option)

[Link to class](https://github.com/spring-projects/spring-statemachine/blob/1.2.5.RELEASE/spring-statemachine-core/src/main/java/org/springframework/statemachine/support/AbstractStateMachine.java#L591)

### PR DESCRIPTION
**Pull request by jvalkeal**
- Resetting a machine with null context currently
  just resets to initial state and clear extended
  state variables. Enhance also to clear machine id
  which as a user level variable may cause some other
  trouble. There are no internal functionality for this
  id as it's meant for user consumption.
- Fixes #390

### CODE REVIEW

**Comment by heffebaycay** [comment]
> There are no internal functionality for this id as it's meant for user consumption.

@jvalkeal Well except, there is... I think. For instance `OnStateEntry` annotated listeners are stored within a cache under a key derived from the state machine name or id (depending on how `WithStateMachine`annotation is used). The cache is built in `StateMachineHandlerCallHelper#afterPropertiesSet`

Short story on how I noticed this:
After migrating to 1.2.6.RELEASE, I noticed my `OnStateEntry` listeners were not called anymore. I'm using Spring State Machine with a custom implementation of `StateMachinePersist` that stores the StateMachineContext in a database table. The instances of StateMachine for a given workflow are all using the same machine name & id (in hindsight, that was a mistake). Whenever the program needs a State Machine, it calls `StateMachinePersister#restore` regardless of whether a context already exists.
In SSM 1.2.6.RELEASE, a call to restore() with a non-existing context instantiates a StateMachine with a null id... which means that later on the listeners will not get called because they are stored under a different cache key.

Using `WithStateMachine` with the `name` parameter (set to the state machine bean name) instead of the `id` parameter fixed the issue for me.

tl;dr: The id parameter of a State Machine is used internally and setting a bad value to it may lead to unexpected behaviors :P
